lib/internal/promisify.js

```diff
@@ -0,0 +1,43 @@
+'use strict';
```

Comment:
I agree with many of the comments here. This file is needlessly complicated. Note that resolve only accepts one argument, and does not distinguish between zero arguments and undefined. And the typeof check is no longer necessary since we don't allow passing callbacks to the promisified version.

Here is a version that seems equivalent:

```js
'use strict';

module.exports = promisify;

function promisify(fn, single) {
  Object.defineProperty(inner, 'name', {
    value: fn.name + 'Promisified'
  });
  return inner;

  function inner() {
    return new Promise((resolve, reject) => {
      // this might be slower than array manipulation, but it might not be.
      fn.call(this, ...arguments, single ? resolve : resolver);
    });

    function resolver(err, value) {
      if (err) reject(err);
      resolve(value);
    }
  }
}
```